### PR TITLE
fix(core-backend): skip `undefined`/`null` bindings in `queryEntityIds` and `ViewStore` selector queries

### PR DESCRIPTION
**Pull request by anmolshres98**
PR to fix `queryEntityIds` throwing on undefined/null bindings

#9463 (["Migrate hollow public/beta ECSQL APIs off withPreparedStatement to withQueryReader"](https://github.com/iTwin/itwinjs-core/pull/9463), commit 1ec1071e8c) changed `IModelDb.queryEntityIds`'s binding semantics:

- **Before**: `ECSqlStatement.bindValues` silently skipped `undefined`/`null` binding entries.
- **After**: `QueryBinder.from(params.bindings)` binds `undefined`/`null` as `NULL`, which throws (`BE_SQLITE_ERROR: Failed to bind parameters: fail to find param index for named param '...'`) if that named parameter isn't actually referenced in the ECSQL statement

`ViewStore.querySelectorValues` has the same latent bug, but the error there is caught and logged, silently returning `[]` instead of crashing.

### Fix

Added `QueryBinder.fromSkippingNullish` (`@internal`, `core/common/src/ConcurrentQuery.ts`) that restores the legacy `bindValues` skip semantics — entries with `undefined`/`null` values are omitted from binding instead of bound as `NULL`. `QueryBinder.from` itself is unchanged (no public API change). Both `queryEntityIds` (`core/backend/src/IModelDb.ts`) and `ViewStore.querySelectorValues` (`core/backend/src/ViewStore.ts`) now use the new helper. Being `@internal` and shared from `core-common`, this can be reused by other packages affected by similar migrations.

w/ help from AnmolDroid🤖

### CODE REVIEW

File: core/common/src/ConcurrentQuery.ts [hunk 1]
```diff
@@ -664,28 +664,52 @@ export class QueryBinder {
...
+   */
+  public static from(args: any[] | object | undefined): QueryBinder {
+    return this.fromImpl(args, false);
+  }
+
+  /**
+   * Same as [[QueryBinder.from]], except entries whose value is `undefined` or `null` are skipped instead of bound as NULL,
+   * matching the legacy `ECSqlStatement.bindValues` semantics. For positional arrays, skipped positions are left unbound
+   * and later positions keep their 1-based index.
+   * @internal
```
Comment (hl662) [inline core/common/src/ConcurrentQuery.ts:707]: Interesting, why not make it available to consumers?

Comment (anmolshres98) [reply core/common/src/ConcurrentQuery.ts:707]: I would rather keep new stuff that is not essential for features or fixes to internal. seems better to be conservative about this imo. If other users need to do this, they can implement it themselves too. Thoughts on that?

Comment (aruniverse) [reply core/common/src/ConcurrentQuery.ts:707]: Lets keep it internal, if others need it we can expose this as beta

File: core/common/src/ConcurrentQuery.ts [hunk 2]
```diff
@@ -664,28 +664,52 @@ export class QueryBinder {
...
+   */
+  public static fromSkippingNullish(args: any[] | object | undefined): QueryBinder {
```
Comment (aruniverse) [inline core/common/src/ConcurrentQuery.ts:709]: `QueryBinder` is a public `core-common` class, so adding this `@internal` static member and using it from `core-backend` through the public package barrel creates a cross-package internal contract on a public API. Per this repository's internal-boundary convention, put the backend-only factory/helper in `core/common/src/internal/cross-package.ts` (with a comment naming `core-backend` as its consumer) and have the backend call that surface instead. That keeps `QueryBinder`'s public API unchanged.

**Review by aruniverse — CHANGES_REQUESTED** [review]
One blocking module-boundary concern; the behavioral implementation otherwise matches the legacy binding semantics. CI is green; this review is based on the PR head shown above.

File: core/common/src/ConcurrentQuery.ts (hunk 2 above)
Comment (aruniverse) [reply core/common/src/ConcurrentQuery.ts:709]: Correcting myself here after a closer, evidence-based re-check — this is **not** a boundary violation, and I'm retracting this as a blocker.

`src/internal/cross-package.ts` curates whole modules that live under `src/internal/` and are otherwise excluded from the package's main public barrel (e.g. this package's own `Snapping`, `RenderMesh`, `PackedFeatureTable`; `core-backend`'s `NativePlatform`/`Symbols`). `ConcurrentQuery.ts` isn't one of those — it's a normal top-level `src/` file already exported through the main `core-common` barrel, so there's no cross-package routing mechanism to detour through here.

Member-level `@internal` tagging on an otherwise-`@public` class is an established, heavily used pattern in this exact API report (371+ occurrences), e.g. `BaseMapLayerSettings.fromBackgroundMapProps`, `CodeSpec.createFromJson`, `Code.equalCodes` — all consumed cross-package through the normal barrel today. `fromSkippingNullish` follows that same precedent, and keeping it colocated on `QueryBinder` (reusing the private `fromImpl`/`bind` dispatch) avoids duplicating that logic elsewhere. `@itwin/no-internal-barrel-imports` is an intra-package cyclic-barrel guard and doesn't apply here either.

No change needed on this point. Sorry for the noise.

File: core/backend/src/test/imodel/IModel.test.ts [hunk 3]
```diff
@@ -1393,6 +1393,30 @@ describe("iModel", () => {
...
+    const parent: string | undefined = undefined;
+    const where = "codevalue is null";
+    const withUnusedUndefined = imodel2.queryEntityIds({ from: "generic.PhysicalObject", where, bindings: { parent } });
+    assert.deepEqual(withUnusedUndefined, baseline);
+
+    const withUnusedNull = imodel2.queryEntityIds({ from: "generic.PhysicalObject", where: "codevalue is null", bindings: { parent: null } });
+    assert.deepEqual(withUnusedNull, baseline);
+
+    const someId = imodel2.queryEntityIds({ from: "bis.element", where: "CodeValue IS NOT NULL", limit: 1 }).values().next().value as string;
+    const codeValue = imodel2.elements.getElement(someId).code.value;
```
Comment (aruniverse) [inline core/backend/src/test/imodel/IModel.test.ts:1412]: Minor/non-blocking: `someId` comes from `.values().next().value` with no guard, so if `imodel2` ever had zero `bis.element` rows with a non-null `CodeValue`, this would throw an unrelated, confusing error out of `elements.getElement(undefined)` rather than a clear assertion failure. Given `imodel2` is a long-lived, heavily reused fixture elsewhere in this file, real risk is low — an `assert.isDefined(someId)` before using it would make a future failure mode self-explanatory, but I wouldn't block on this.

File: core/backend/src/test/standalone/ViewStoreDb.test.ts [hunk 4]
```diff
@@ -330,4 +332,33 @@ describe("ViewStore", function (this: Suite) {
 
     vs1.vacuum();
   });
+
+  it("selector queries should skip undefined and null bindings", async () => {
+    const dbName = join(KnownTestLocations.outputDir, "viewStoreBindings.db");
```
Comment (aruniverse) [inline core/backend/src/test/standalone/ViewStoreDb.test.ts:338]: Minor/non-blocking: `viewStoreBindings.db` is created without first removing a leftover file from a prior interrupted run. Low risk of cross-run flakiness given the existing `try/finally` cleanup, but an upfront `IModelJsFs.removeSync` (or similar) guard would make repeated local runs more robust if this pattern isn't already handled by test-output cleanup elsewhere.

**Review by aruniverse — APPROVED** [review]
Re-ran an independent adversarial review round (fresh-context, no reliance on my earlier pass) covering correctness, test realism, and API/module-boundary discipline against this same commit.

**Retracting my prior "Request changes" review.** The module-boundary concern I raised on `ConcurrentQuery.ts:709` doesn't hold up under closer inspection — replied inline with the evidence. `src/internal/cross-package.ts` curates whole modules excluded from a package's main barrel; it isn't the mechanism for member-level `@internal` tags on already-`@public` classes, which is an established pattern elsewhere in this API report (e.g. `BaseMapLayerSettings.fromBackgroundMapProps`, `CodeSpec.createFromJson`). No boundary violation here.

Behaviorally, `fromSkippingNullish` correctly preserves the legacy `ECSqlStatement.bindValues` skip semantics for both named and positional bindings (positional index is captured before the skip check, so later real values keep their correct 1-based index), and both real call sites (`IModelDb.queryEntityIds`, `ViewStore` selector queries) are exercised through the real native binding path in tests, not just the `QueryBinder` unit level.

Two small, non-blocking test-robustness notes below — neither is required before merging.

**Comment by aruniverse** [comment]
I've approved, but please do not merge until the rest of the ECDb that I have tagged reviews.
cc @khanaffan @rschili

**Review by markschlosseratbentley — APPROVED** [review]

**Review by rschili — APPROVED** [review]

**Comment by khanaffan** [comment]
@diegoalexdiaz had issue similar to this while migrating `bindIdSet([0x1,undefined, 0x2])` where there is undefined in IdSet cause failure or crash. 

Can you add unit test to cover that as well to make sure it work the same with `withQueryReader()` as it does with `ECSqlReader`

**Comment by anmolshres98** [comment]
> @diegoalexdiaz had issue similar to this while migrating `bindIdSet([0x1,undefined, 0x2])` where there is undefined in IdSet cause failure or crash.
> 
> Can you add unit test to cover that as well to make sure it work the same with `withQueryReader()` as it does with `ECSqlReader`

after a deeper validation run imo this is a separate bug from what this PR fixes, so I filed a separate issue for it: https://github.com/iTwin/itwinjs-core/issues/9530

this PR only fixes the regression from #9463 in `fromSkippingNullish`/`queryEntityIds`/`ViewStore`. the crash you're describing comes from `bindIdSet` itself, a different code path with no overlap in this diff. I'll have the tests coverage for the same in the other pr too

**Review by khanaffan — APPROVED** [review]
